ngraph/core/reference/include/ngraph/runtime/reference/adaptive_avg_pool.hpp

```diff
@@ -0,0 +1,169 @@
+                            size_t w_start = window_start(j, w_in, w_out);
+                            size_t w_end = window_end(j, w_in, w_out);
+                            T result = 0;
+                            for (size_t n = h_start; n < h_end; n++)
+                                result = std::accumulate(
+                                    arg + n * w_in + w_start, arg + n * w_in + w_end, result);
+                            out[i * w_out + j] = avg(result, (w_end - w_start) * (h_end - h_start));
+                        }
+                    }
+                }
+                template <typename T>
+                void adaptive_avg_pool_3d(const T* arg,
+                                          T* out,
+                                          size_t d_in,
+                                          size_t d_out,
+                                          size_t h_in,
+                                          size_t h_out,
+                                          size_t w_in,
+                                          size_t w_out)
+                {
+                    for (size_t i = 0; i < d_out; i++)
+                    {
+                        size_t d_start = window_start(i, d_in, d_out);
+                        size_t d_end = window_end(i, d_in, d_out);
+                        for (size_t j = 0; j < h_out; j++)
+                        {
+                            size_t h_start = window_start(j, h_in, h_out);
+                            size_t h_end = window_end(j, h_in, h_out);
+                            for (size_t k = 0; k < w_out; k++)
+                            {
+                                size_t w_start = window_start(k, w_in, w_out);
+                                size_t w_end = window_end(k, w_in, w_out);
+                                T result = 0;
+                                for (size_t n = d_start; n < d_end; n++)
+                                    for (size_t m = h_start; m < h_end; m++)
+                                    {
+                                        auto pos = arg + n * h_in * w_in + m * w_in;
+                                        result =
+                                            std::accumulate(pos + w_start, pos + w_end, result);
+                                    }
+                                out[i * h_out * w_out + j * w_out + k] =
+                                    avg(result,
+                                        (d_end - d_start) * (w_end - w_start) * (h_end - h_start));
+                            }
+                        }
+                    }
+                }
+            } // namespace adaptive_pool
+            template <typename T>
+            void adaptive_avg_pool(const T* arg,
+                                   T* out,
+                                   const Shape& arg_shape,
+                                   const Shape& out_shape)
+            {
+                size_t channel_size = 1;
```

Comment:
Do we need or not checks:
- `arg_shape.size() == out_shape.size()`
- `2 < arg_shape.size() && arg_shape.size() < 6`